### ISSUE TITLE
Think about restructuring templates

### ISSUE THREAD

**Issue by ralsina** [getnikola/nikola#1814]
One thing I noticed when doing the theme tutorial. At one point I need to change a function that's in [base_header.tmpl](https://github.com/getnikola/nikola/blob/master/nikola/data/themes/base/templates/base_header.tmpl), say, ``html_navigation_links``.

Currently, that means I need to copy that whole file, and change that function. That has ... problems:

* I have to find that file (not horribly hard now)
* I need to find that function in a sea of code (scary for newbies)
* I have to bring a bunch of code into my theme that I *don't* care about (mo' code, mo' problems)
* When the parent theme is updated, I am not getting the fixes.

So, basically, we are making derived themes large and difficult.

So, what about having a one-function-per-file policy? Or something else that allows us to make this easier. It doesn't even need to be incompatible, we can keep ``base_helper.tmpl`` and just have it be a namespace that imports a bunch of other stuff.

**Comment by Kwpolska** [comment]
> (scary for newbies)

Even notepad.exe has a Find function.

> one-function-per-file

This sounds really messy.

---

My proposal is: make parent templates accessible in templates.  This should be doable if we ask nicely, and we could just use template inheritance/extending with that:


```mako
## -*- coding: utf-8 -*-
<%inherit file="parent/base/base_header.tmpl"/>
<%def name="html_navigation_links()">
Custom code goes here.
</%def>
## end of file
```

```jinja
{# -*- coding: utf-8 -*- #}
{% extends 'parent/base-jinja/base_header.tmpl' %}
{% macro html_navigation_links() %}
Custom code goes here.
{% endmacro %}
{# end of file #}
```

Mako: no idea (might need some extra code…).  Jinja2: [probably this](http://jinja.pocoo.org/docs/dev/api/#jinja2.PrefixLoader) (assuming we can do empty prefixes w/o leading slashes appearing all over the place).

**Comment by schettino72** [comment]
> So, what about having a one-function-per-file policy?
:+1: 

It was so hard to keep my small customizations on my template at every update that I slowly gave up...

Once I thought about having some patches instead of inheriting from a base template. 
I guess for small changes patches might be easier to manage after an upgrade.
Not sure it would be worth the effort, just throwing out this idea...

**Comment by damianavila** [comment]
> one-function-per-file
> This sounds really messy.

I second @Kwpolska idea here...

**Comment by ralsina** [comment]
How about exposing the parent of each theme in the template locator?

That way, if you want to create themeA/templates/foo.tmpl that inherits themeB/templates/foo.tmpl (which you can't now!) you just do it.

Makes sense?

**Comment by ralsina** [comment]
I have proposed a solution that avoids duplicating code at no cost :-)

https://github.com/getnikola/nikola/compare/idea-for-1814?expand=1

I only simplified one template but using meld a whole lot can go away.

**Comment by ralsina** [comment]
Merged the idea mentioned in previous branch.
